lib/commands/touch.js

```diff
@@ -1,9 +1,99 @@
+  for (const command in actions) {
+    if (actions[command].type === 'key') {
+      this.log.info('Mapping keyboard actions');
+      return await performKeyboardCommands.bind(this)(actions[command].actions);
+    } else if (actions[command].type === 'pointer' && actions[command].parameters.pointerType === 'mouse') {
+      this.log.info('Mapping mouse actions');
+      return await performMouseCommands.bind(this)(actions[command].actions);
+    } else {
+      this.log.info('Going default actions route');
+      return await this.winAppDriver.sendCommand('/actions', 'POST', {actions});
+    }
+  }
 };
 
+/*
+  Supported actions:
+  -> sendKeys(CharSequence... keys)
+  -> sendKeysInTicks(CharSequence... keys)
+*/
+// TODO: remove once WAD offers W3C for keyboard input actions
+async function performKeyboardCommands(actions) {
+  const keysFormatted = formatKeyboardInput(actions);
+  return await this.winAppDriver.sendCommand('/keys', 'POST', {'value': keysFormatted });
+}
+
+/*
+  Supported actions:
+  -> click()
+  -> click(WebElement target)
+  -> contextClick()
+  -> contextClick(WebElement target)
+  -> release()
+  -> release(WebElement target)
+  -> clickAndHold()
+  -> clickAndHold(WebElement target)
+  -> doubleClick()
+  -> doubleClick(WebElement target)
+  -> moveToElement(WebElement target)
+  -> moveToElement(WebElement target, int xOffset, int yOffset)
+  -> dragAndDrop(WebElement source, WebElement target)
+  -> pause()
+
+*/
+// TODO: remove once WAD offers W3C support for pointer actions of type mouse
+async function performMouseCommands(actions) {
+  let result;
+  for (const action in actions) {
+    switch (actions[action].type) {
+      case 'pointerMove':
+        this.log.info(JSON.stringify(actions[action]));
+        this.log.info('Done');
+        result = await this.winAppDriver.sendCommand('/moveto', 'POST', getExpectedFormatForMouseMovement(actions[action]));
+        break;
+      case 'pointerDown':
+        result = await this.winAppDriver.sendCommand('/buttondown', 'POST', {'button': actions[action].button });
+        break;
+      case 'pointerUp':
+        result = await this.winAppDriver.sendCommand('/buttonup', 'POST', {'button': actions[action].button });
+        break;
+      case 'pause':
+        await sleep(actions[action].duration);
+        break;
+      default:
+        this.log.info('No mouse action mapped for type: ${JSON.stringify(actions[action].type)}, moving to next action');
+    }
+  };
+  return result;
+}
+
+function formatKeyboardInput(actions) {
+  const keys = new Array();
+  actions.forEach((action) => {
+    if (action.type === 'keyDown') {
+      keys.push(action.value);
+    }
+  });
+  return [keys.join('')];
+}
+
+function getExpectedFormatForMouseMovement(action) {
+  const elementId = util.unwrapElement(action.origin);
+  if (action.x !== 0 && action.y !== 0) {
```

Comment:
Could you please link me to the place in w3c actions spec where the format of mouse gesture is described?

Reply:
Can't really point you to something specific. There's this part https://www.w3.org/TR/pointerevents/#the-pointermove-event but it doesn't show the format of the json for the action which is what interests us.

Why I added that logic:
Basically this is how the action for pointerMove looks
This is json resulted from calling pinterMove(Element) -> it automatically sets the x and y offset to 0
{"duration":100,"x":0,"y":0,"type":"pointerMove","origin":{"ELEMENT":"7.1824.8897958","element-6066-11e4-a52e-4f735466cecf":"7.1824.8897958"}}
If we se a specific offset for the movement then it will be reflected in the x and y field.

The trouble is that the WAD endpoint for /moveTo which is the equivalent of moving to an element has some extra logic behind it so that when only the element is provide it knows to move the pointer position to the center of the element.

So if we provide just the elementID -> Moves pointer to center of element (calculates and adds offset behind the scenes from what I saw)
If we provide element with offsets -> Moves it to element corner + offsets

So when we call the pointerMove(Element) method on appium side we expect it to move to middle of element.
That's why when I call the WAD endpoint I set the offsets x and y only if they have a value. If it's 0 for both of them I just send the element as parameter so it knows to move to the center automatically

Sorry for the long explanation

Reply:
Is this step-by-step guide helpful: https://www.w3.org/TR/webdriver/#dfn-process-a-pointer-move-action ?